src/core/tsi/s2a/handshaker/s2a_tsi_handshaker.cc

```diff
@@ -16,34 +16,327 @@
+static const tsi_handshaker_vtable handshaker_vtable = {
+    nullptr,         nullptr,
+    nullptr,         nullptr,
+    nullptr,         handshaker_destroy,
+    handshaker_next, handshaker_shutdown};
+
 tsi_result s2a_tsi_handshaker_create(
     const grpc_s2a_credentials_options* options, const char* target_name,
     const char* handshaker_service_url, bool is_client,
     grpc_pollset_set* interested_parties, tsi_handshaker** self,
     char** error_details) {
+  if (options == nullptr || (is_client && target_name == nullptr) ||
+      handshaker_service_url == nullptr || self == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerNullptrArguments);
+    return TSI_INVALID_ARGUMENT;
+  }
+  s2a_tsi_handshaker* handshaker =
+      static_cast<s2a_tsi_handshaker*>(gpr_zalloc(sizeof(s2a_tsi_handshaker)));
+  gpr_mu_init(&(handshaker->mu));
+  handshaker->client = nullptr;
+  handshaker->is_client = is_client;
+  handshaker->has_sent_start_message = false;
+  handshaker->target_name = (target_name == nullptr)
+                                ? grpc_empty_slice()
+                                : grpc_slice_from_static_string(target_name);
+  handshaker->interested_parties = interested_parties;
+  handshaker->has_created_handshaker_client = false;
+  handshaker->handshaker_service_url = gpr_strdup(handshaker_service_url);
+  // TODO(mattstev): this API is exposed in a PR that is not yet merged.
+  // handshaker->options = grpc_s2a_credentials_options_copy(options);
+  handshaker->base.vtable = &handshaker_vtable;
+
+  *self = &(handshaker->base);
+  return TSI_OK;
+}
+
+static tsi_result s2a_handshaker_result_extract_peer(
+    const tsi_handshaker_result* self, tsi_peer* peer) {
+  // TODO(mattstev): implement.
   return TSI_UNIMPLEMENTED;
 }
 
+static tsi_result s2a_handshaker_result_create_zero_copy_grpc_protector(
+    const tsi_handshaker_result* self, size_t* max_output_protected_frame_size,
+    tsi_zero_copy_grpc_protector** protector) {
+  // TODO(mattstev): the implementation is blocked because the necessary API's
+  // are exposed in a PR that is not yet merged.
+  return TSI_UNIMPLEMENTED;
+}
+
+static tsi_result s2a_handshaker_result_get_unused_bytes(
+    const tsi_handshaker_result* self, const unsigned char** bytes,
+    size_t* bytes_size) {
+  if (self == nullptr || bytes == nullptr || bytes_size == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultUnusedBytesNullptr);
+    return TSI_INVALID_ARGUMENT;
+  }
+  s2a_tsi_handshaker_result* result =
+      reinterpret_cast<s2a_tsi_handshaker_result*>(
+          const_cast<tsi_handshaker_result*>(self));
+  *bytes = result->unused_bytes;
+  *bytes_size = result->unused_bytes_size;
+  return TSI_OK;
+}
+
+static void s2a_handshaker_result_destroy(tsi_handshaker_result* self) {
+  if (self == nullptr) {
+    return;
+  }
+  s2a_tsi_handshaker_result* result =
+      reinterpret_cast<s2a_tsi_handshaker_result*>(self);
+  // TODO(mattstev): do I need to add nullptr checks?
```

Comment:
I do not think you need. `gpr_free` will do a `nullptr` check and will be a no-op if it is `nullptr`.

Reply:
Sorry forgot to remove my comment! Thanks!